feat(Observable): initially create Observable.generate factory

Igorbek opened the pull request:
**Description:**

The PR adds static `Observable.generate/GenerateObservable.create` method with implementation aligned with RxJS4.

The signature supports passing parameters and a single options:

```ts
  static create<T, S>(
    initialState: S,
    condition: ConditionFunc<S>,
    iterate: IterateFunc<S>,
    resultSelector: ResultFunc<S, T>,
    scheduler?: Scheduler): Observable<T>
  static create<S>(
    initialState: S,
    condition: ConditionFunc<S>,
    iterate: IterateFunc<S>,
    scheduler?: Scheduler): Observable<S>
  static create<T, S>(
    options: GenerateOptions<T, S>): Observable<T>

```

**Related issue:**

#1482 

_This is my first PR here, hi everybody_ :)

kwonoj commented on src/Rx.ts:
```diff
@@ -20,6 +20,7 @@ import './add/observable/fromArray';
 import './add/observable/fromEvent';
 import './add/observable/fromEventPattern';
 import './add/observable/fromPromise';
+import './add/observable/generate';
```
if this is going under `Rx`, it may need to be also under `kitchenSink` as well. (`kitchSink` contains all of `Rx` export)

Igorbek replied:
I didn't know what should be included into `kitchenSink`. I'll include.

david-driscoll replied:
@kwonoj kitechen sink imports all of Rx by default, so if it is part of rx, kitchen sink will see it too.

kwonoj replied:
Oops, I missed that. Sorry for bringing confusion.

kwonoj commented on src/observable/GenerateObservable.ts:
```diff
@@ -0,0 +1,129 @@
+  }
+
+  static create<T, S>(
+    initialState: S,
```
it's just style - would you able to flatten signature into single line similar to [`combineLatest`](https://github.com/ReactiveX/rxjs/blob/master/src/operator/combineLatest.ts#L45-L60) ? it's not strict rules but generally follows that way for convenience comparison between signatures.

kwonoj commented on src/observable/GenerateObservable.ts:
```diff
@@ -0,0 +1,129 @@
+        <Scheduler>resultSelectorOrObservable);
+    }
+
+    return new GenerateObservable<T, S>(
```
does test case covers this case too?

Igorbek replied:
I've added a new one that tests this case.

kwonoj commented on src/observable/GenerateObservable.ts:
```diff
@@ -0,0 +1,129 @@
+        params: {
+          subscriber,
+          iterate: this.iterate,
+          condition: this.condition,
```
you can declare const variable `const condition = this.condition; const resultSelector = this....` for variable being used more than once in function scope.

kwonoj commented on src/observable/GenerateObservable.ts:
```diff
@@ -0,0 +1,129 @@
+    iterate: IterateFunc<S>;
+    resultSelector: ResultFunc<S, T>;
+  };
+}
```
what about flatten `SchedulerState` similar to

```js
SchedulerState<T, S> {
  state: S;
  subscriber: Subscriber<T>;
  condition?: ConditionFunc<S>;
  iterate: IterateFunc<S>;
  resultSelector: ResultFunc<S, T>;
}
```
? then pass mutated object at line 127.

blesh commented:
@Igorbek this PR needs to be flattened and rebased.

Igorbek commented:
Sure.

Igorbek commented on spec/observables/generate-spec.ts:
```diff
@@ -0,0 +1,77 @@
+    const expected = '(12345|)';
+
+    expectObservable(source).toBe(expected);
+  });*/
```
to enable this test need to fix #1543 first

kwonoj replied:
I've checked in https://github.com/ReactiveX/rxjs/pull/1544 prior to this and believe this is now unblocked - would you mind to update this one?

Igorbek commented:
@blesh, it's been flattened and rebased.

blesh commented:
@Igorbek we might need to hold this PR until @kwonoj's #1545 lands, or otherwise coordinate with @kwonoj. I don't think there will be _huge_ changes to the testing code, but there would be some changes for sure.

@kwonoj, what do you think?

kwonoj commented:
I'm about to check in #1545 now (locally running tests for verification once again) so it'll be soon - would like to ask same way to hold this bit. Apologize to @Igorbek might need one additional rebase.

Igorbek commented:
Sure, no problem.

Igorbek commented:
Ok, now it's been fixed, flattened and rebased again.

staltz commented:
Sorry for reviewing this late in the process, and thank you for the effort so far.

I'd like to remind about the criteria for merging an operator (this was shown as a [PR comment template](https://raw.githubusercontent.com/ReactiveX/rxjs/master/.github/PULL_REQUEST_TEMPLATE.md) when a PR is opened), so there are still a few things missing until this operator is "complete", namely:

- [ ] If possible, write a `asDiagram` test case too, for PNG diagram generation purposes
- [ ] The operator must be documented in JSDoc style in the implementation file, including also the PNG marble diagram image
- [ ] The operator should be listed in `doc/operators.md` in a category of operators
- [ ] It should also be inserted in the operator decision tree file `doc/decision-tree-widget/tree.yml`

If you need a reference in how to do those, I recommend looking at 
- `src/MiscJSDoc.ts`
- `src/observable/EmptyObservable.ts`
- https://github.com/ReactiveX/rxjs/blob/7efcdc99973d496b70702eb6a40dd384a2e702c4/spec/Observable-spec.ts#L492
- https://github.com/ReactiveX/rxjs/blob/7efcdc99973d496b70702eb6a40dd384a2e702c4/doc/decision-tree-widget/tree.yml#L349

Igorbek commented:
Ok, I've just returned from Build and so now will probably have more time this week to complete this journey.

Igorbek commented on doc/operators.md:
```diff
@@ -120,6 +120,7 @@ There are operators for different purposes, and they may be categorized as: crea
 - [`fromEvent`](../class/es6/Observable.js~Observable.html#static-method-fromEvent)
 - [`fromEventPattern`](../class/es6/Observable.js~Observable.html#static-method-fromEventPattern)
 - [`fromPromise`](../class/es6/Observable.js~Observable.html#static-method-fromPromise)
+- `generate`
```
@staltz should I make it a link by following the pattern around?

staltz replied:
Yes, a link is nice :+1:

Igorbek replied:
done

Igorbek commented:
Ok, looks like I'm done. That's what was changed:
- added an overload that accepts an 'options' object without `resultSelector` but still properly typed with state as a result type
- all calls to user functions (`iterate`, `condition` and `resultSelector`) are now in `try-catch` errors are emitted to subscriber's error channel
- reduced unnecessary calls to `condition` function
- added more tests to cover scenarios with/without scheduler specified
- added JSDoc to all exported functions and option fields
- the operator added to the list in doc and to decision tree

@staltz please review and see my question about doc link

staltz commented on spec/observables/generate-spec.ts:
```diff
@@ -2,36 +2,43 @@ import * as Rx from '../../dist/cjs/Rx.KitchenSink';
   });
-  it('should produce all values synchronously', () => {
+
+  asDiagram('generate(1, x => x < 3, x => x + 1)')
```
Usually we have only one of these `asDiagram` tests, per file. This third one is the most interesting for a diagram. Can you make the 1st and 2nd test revert back to normal `it()` test?

staltz commented on doc/decision-tree-widget/tree.yml:
```diff
@@ -349,6 +349,9 @@ children:
       - label: using custom logic
         children:
         - label: Observable.create
+      - label: using state machine
```
`using a state machine similar to a for loop`

staltz commented on src/observable/GenerateObservable.ts:
```diff
@@ -37,16 +60,73 @@ export class GenerateObservable<T, S> extends Observable<T> {
+   * @param {Function} resultSelector Selector function for results produced in the sequence.
+   * @param {Scheduler} scheduler Scheduler on which to run the generator loop. If not provided, defaults to emit immediately.
+   * @returns {Observable} The generated sequence.
+   */
```
There is a lot of details missing here still, if you compare to other operator JSDocs. Would you prefer to update this PR or do you prefer that I write out the missing details in another PR?

Igorbek replied:
I'm ok with either case, but would need some advises to what exactly is messing and what to add. So create separate PR if you like.

staltz replied:
The structure we use in other operators is (in this order):
- 2 line short technical description of the operator.
- 2 line short *informal* description (with plain English, for dummies) of the operator.
- `<img>` with the marble diagram
- Long technical description of the operator, in details and explaining all the optional arguments too.
- Examples (yours are good already :+1:)
- `@see` linking to similar operators. E.g. I would link to `from` and `create`.
- Params (yours is good, but `Mixed` is not a valid type)

It's easy for me to do a separate PR for this, if you prefer.

Igorbek commented:
@staltz Please go ahead with your separate PR for the docs. I've slightly changed the docs, so I hope it'll help you. The _informal_ part and _long technical description_ are still missing. I appreciate your assistance.

staltz commented:
Nice, thanks

blesh commented:
@staltz ... this looks okay from my review, but I'm leaving it up to you since it seems like you're waiting for something docs related.

blesh commented:
Okay.. this LGTM. Last thing would be to flatten this into one commit. @Igorbek can you flatten this? The commit message should be `feat(Observable.generate): add generate static creation method` with some description of the work done. This is so it shows up cleanly in the CHANGELOG.md at publish time.

I'm publishing in a few minutes, so this will probably land in next week's release.

Igorbek commented:
Done.

staltz commented:
LGTM, thanks Igor

blesh commented:
Merged with c03434c... thanks @Igorbek

Igorbek commented:
@blesh just curious, why don't you use normal merge and use this strange cherry picking?

staltz commented:
@Igorbek I think it's because of https://github.com/ReactiveX/rxjs/blob/master/doc/maintainer-guidelines.md
